Add: new blog pycon talk - python packaging (#405)
pyOpenSci/pyopensci.github.io: _posts/2024-05-31-friends-dont-let-friends-package-alone-pyconus-24.md

---
layout: single
title: "Friends don't let friends package alone- my Python packaging talk at PyCon 2024"
excerpt: "."
author: "Leah Wasser"
permalink: /blog/python-packaging-friends-dont-let-friends-package-alone.html
header:
  overlay_image: images/blog/2024/may/pycon-us-2024.png
  overlay_filter: 0.6
categories:
  - blog-post
  - community
classes: wide
toc: true
comments: true
---

## <i class="fa-solid fa-heart"></i>  Python packaging and community: my first talk at PyCon US 2024

This year I gave my first talk in the [PyCon US](https://us.pycon.org/2024/) main track! For context, PyCon US is a big meeting - with over 2,700 registrations this year - a record for the event! My talk was about my experiences navigating the packaging ecosystem from a few perspectives:

* teaching scientists how to share their code,

> * a maintainer of a smaller python package - stravalib and

Reviewer: ```suggestion
* as a maintainer of a smaller Python package - stravalib and
```

> * founder and executive director of a non profit organization - pyOpenSci that strives to support scientists in tackling the worlds greatest challenges using open software and workflows.

Reviewer: ```suggestion
* Founder and Executive Director of a non profit organization - pyOpenSci - which strives to support scientists in tackling the world's greatest challenges using open software and workflows.
```

The irony of these three  different perspectives is that the challenges that users face in each category are similar! Below I explore my experiences pulling together the talk.

### My early experiences with Python packaging

I started thinking about Python packaging in 2018 during our [early pyOpenSci
community meetings](https://www.pyopensci.org/handbook/reference/index.html). In these meetings, we were trying to create an [open peer
review process for scientific Python software](https://www.pyopensci.org/about-peer-review/index.html). A process that is running today

> with 17 scientific Python packages currently in some state of review as write

Reviewer: ```suggestion
with 17 scientific Python packages currently [in some state of review](https://github.com/pyOpenSci/software-submission/issues) as write
```

this post.

> To me, as someone who was teaching Python to beginners and dabbling in creating

Reviewer: ```suggestion
As someone who was teaching Python to beginners and dabbling in creating
```

and maintaining packages to support teaching, packaging was this somewhat
undefined space. I wasn’t sure where to begin or what sources were truly
authoritative. Generally, my approach to packaging was copying the
infrastructure that other package maintainers I trusted used—mainly GeoPandas.

*I am a spatial data nerd at heart.*

<figure>
  <img src="{{ site.baseurl }}/images/blog/2024/june/pycon-python-packaging-confusion.png" alt="Image that says  have you ever felt confused about python packaging. Below is the pyopensci flower logo. And the graphic shows a bunch of different cartoon like hands being raised." style="width: 90%;" />
  <figcaption> Every time I ask this question, a majority of people in the room raise their hand.</figcaption>
</figure>

But as I thought about creating pyOpenSci's open peer review process, I knew
scientists would need a process that:

* Enforced “good enough” practices in scientific Python packaging

> * Was not overly binding but also set users up for success

Reviewer: ```suggestion
* Was not overly-binding, but also set users up for success
```

I also knew that:

> * We needed to create some sort of packaging guide that set up expectations for our pyOpenSci open peer review of software process. And that

Reviewer: ```suggestion
* We needed to create some sort of packaging guide that set up expectations for our pyOpenSci open peer review of software process, and that
```

> * I wasn't the lead expert in the room. Thus, needed people from across the (scientific) Python ecosystem and beyond to build consensus and define what

Reviewer: ```suggestion
* I wasn't the lead expert in the room. Thus, we needed people from across the (scientific) Python ecosystem and beyond to build consensus and define what
```

> were “good enough” practices.

Reviewer: ```suggestion 
“good enough” practices were.
```

> But how could pyOpenSci in these early years, as a not-yet-created community empower people to

Reviewer: ```suggestion
But how could pyOpenSci in these early years, as a not-yet-created community, empower people to
```

> navigate a tricky if not complex ecosystem empowered by choice?

Reviewer: ```suggestion
navigate a tricky, if not complex, ecosystem empowered by choice?
```

**Choice is great but it's not necessarily beginner friendly.** Choice leaves too much room for uncertainty as a beginner. And uncertainty just makes learning more stressful.

As my mentor and colleague Carol likes to say:

> choice is an opportunity for someone to make the wrong decision

So if you think about that "opportunity to make a wrong decision", imagine what happens if you have multiple choices that lead you down a terrifying path of uncertainty! <i class="fa-solid fa-ghost"></i>

<figure>
  <img src="{{ site.baseurl }}/images/blog/2024/june/pycon-python-packaging-too-many-options.png" alt="Image of a flat coated retriever lying down with a purple ice bag on its head looking sad." style="width: 90%;" />
  <figcaption> Too many choices can lead to frustration for many users that are learning a new technical thing. More biscuits - or early wins - are more empowering for a learner and set them up for success.</figcaption>
</figure>

### The evolution of the pyOpenSci Python packaging guide
Years later, what started out as a generic Python development guide, was
transformed into what’s now the [pyOpenSci Python package guide](https://www.pyopensci.org/python-package-guide/).

> 

Reviewer: The pyOpenSci package guide is a guidebook that:

> * A guidebook that has 58 contributors as I’m writing this post today.

Reviewer: ```suggestion
* 58 contributors as I’m writing this post today.
```

> * A guidebook that has color and graphics and that leads Pythonistas down a

Reviewer: ```suggestion
* color and graphics and that leads Pythonistas down a
```

  singular, opinionated path to Python packaging success.

> * A guidebook that breaks down pieces and processes into their fundamentals and

Reviewer: ```suggestion
* breaks down pieces and processes into their fundamentals and
```

  explains them using digestible words.

The pyOpenSci path is not the only path. It is just a path that works and
enables early wins for scientists and others using Python.

> If you're reading this and you have your own path, that's ok too. But if you

Reviewer: ```suggestion
If you're reading this and you have your own path, that's OK too. But if you
```

want to worry less about decisions and have a definitive way of doing things
that follows modern standards -

we've got you.

### A packaging guidebook created by community

This guidebook was created by a community of people with beginner to expert packaging experience with the goal of building consensus
around a single approach to pure Python packaging that would **yield user success**.

> Together we created a guidebook that acknowledges a suite of fantastic tools across the ecosystem while

Reviewer: ```suggestion
Together we created a guidebook that acknowledges a suite of fantastic tools across the ecosystem, while
```

still leading users down a streamlined path.

<figure>
  <img src="{{ site.baseurl }}/images/blog/2024/june/pyopensci-packaging-guide-overview.png" alt="Image with the pyOpenSci flower logo in the upper right hand corner. The image shows the packaging lifecycle. The graphic shows a high level overview of the elements of a Python package. the inside circle has 5 items - user documentation, code/api, test suite, contributor documentation, project metadata / license / readme. In the middle of the circle is says maintainers and has a small icon with people. on the outside circle there is an arrow and it says infrastructure." style="width: 70%;" />
  <figcaption><i class="fa-solid fa-heart-pulse"></i> The generalized scientific Python package development process. <i class="fa-solid fa-heart-pulse"></i>.</figcaption>
</figure>

## Getting accepted to talk about packaging...  <i class="fa-solid fa-ghost"></i>

> I was thrilled and terrified to be accepted to give a talk on the process of creating the pyOpenSci Python packaging guidebook at pyCon US in Pittsburgh, Pennsylvania. My talk was called:

Reviewer: ```suggestion
I was thrilled and terrified to be accepted to give a talk at [PyCon US](https://us.pycon.org/2024/), in Pittsburgh, Pennsylvania, on the process of creating the pyOpenSci Python packaging guidebook . My talk was called:
```

“Friends don’t let friends package alone”

It was all about the process of navigating the Python packaging ecosystem and
then creating the guide. While you might assume such a talk about be technical,
in reality, it was all about people.

### Me and the child-like imposter inside

> Giving this talk was a big deal for me. Imposter syndrome—the feeling that I

Reviewer: ```suggestion
Giving this talk was a big deal for me. Imposter syndrome—the feeling that I
```

Reviewer: one thing to mention is that the use of hyphens (essentially in place of em-dashes) is inconsistent. I'm not sure that the --- gets converted to an em-dash in Markdown, but for consistency we should be doing either words connected by the dash (syndrome-the) or with spaces (syndrome - the).

with an em-dash, the correct way is to have the dash "touch" the words, without a space. but if we don't get a rendered em-dash (the longest hyphen), then spaces make more sense, as they increase readability.

Author: @kierisi i always get confused with em dashes - i know there are different types of dashes. i don't know when ti use which one. is em dash the longer one so it should always touch the words? if that is the case i can fix throughout locally.

Reviewer: I only know because we ended up doing a deep dive on this when we wrote Data Science in Education Using R! so en dashes are *generally* used for ranges in numbers and dates, and em dashes are a replacement for commas or parentheses. the way I remember it is that an eN dash is for Numbers. in both cases (along with hyphens!) the dash should be touching both words.

Author: ok - that's helpful. It looks like to use an em-dash i'd have to add the unicode value

&#151; 

i can try that? otherwise markdown is not fancy and doesn't know about typography things. Let me see how unicode renders!

Reviewer: sounds good! honestly I think an en-dash is fine, it's more that the style is consistent!

Author: actually just pasting the character in also works - `—` that's much easier.

didn't belong, that I wasn't technical enough to build pyOpenSci — has always been
my reality. As a trained Landscape Architect, I spent most of my college and early graduate years

> practicing perfect lettering, using markers, fancy pens and colored pencils  to bring creative

Reviewer: This should be a blog post and talk!!!

Author: imposter syndrome ? or the literal dozen of tubes with drawings that i can't let go of that i spent weeks on that are in my garage 😆 or both? i think the markers are all dried up.

i was terrible at lettering btw. never been good at handwriting. 😆

Reviewer: ummmm I'd definitely read a post going through some of those tubes of drawings!

landscape visions to life.

Now, as a Pythonista, coder, data scientist, and tech
geek, I’m very aware of how much more others know than I do about lots of
different topics—including packaging.

> I wasn’t sure what it would be like to be on that big pyCon US presentation stage, in that big room full of chairs. Would they be empty? Luckily, I was able to focus on my friends sitting in the front rows

Reviewer: ```suggestion
I wasn’t sure what it would be like to be on that big pyCon US presentation stage, in that big room full of chairs. Would they be empty? Luckily, I was able to focus on my friends sitting in the front rows.
```

Carol, Hugo, Jeremiah, and Inessa - they were all there.

Carol told me to lean into the teacher inside of me. I know that teacher best. My body calmed as I remembered why I was there. To help the community move forward.

And similar to how we created our packaging guide - my friends got me through it.

## The core messages in my talk

> In my talk, I shared my experiences as a maintainer of [Stravalib](https://github.com/stravalib/stravalib). How

Reviewer: ```suggestion
In my talk, I shared my experiences as a maintainer of [Stravalib](https://github.com/stravalib/stravalib), and how
```

challenging it was to figure out the best packaging infrastructure for our team.

> However, that experience of rebuilding stravalib also empowered me to better understand the pain points

Reviewer: stravalib appears as both 'stravalib' and 'Stravalib' throughout the piece

Author: stravalib has no brand but i think we use it lower case throughout except when it begins a sentence so i'll fix locally. thanks!

Author: fixed.

  <figcaption><i class="fa-solid fa-heart-pulse"></i> Even as technical maintainers, the folks on the stravalib team weren't packaging experts. Why would they be? They are focused on writing good usable code (and tests). <i class="fa-solid fa-heart-pulse"></i>.</figcaption>
</figure>

## Fundamental first - How to make a technical ecosystem beginner accessible

How do we collectively support and empower people who don't want to become
packaging gurus to be successful?

My approach to helping beginners tackle hard technical problems has always been
about breaking things down into their simplest forms first.

I thrive on those “ah ha” moments when a tough concept clicks.

And someone gets it.

Sometimes, that someone is me — before I can teach the topic to others!

## So what about packaging with friends?

The essence of my Python packaging talk was the work that I have lead alongside the pyOpenSci community to demystify and help _people_ navigate the complex Python packaging ecosystem.

This work was really about people helping people tackle challenging things.

> Python packaging matters because scientists need open software to create open and reproducible workflows. To build upon each other's work. To tackle the world's hardest challenges more efficiently and effectively.

Reviewer: ```suggestion
Python packaging matters because scientists need open software to create open and reproducible workflows to build upon each other's work, and to tackle the world's hardest challenges more efficiently and effectively.
```

* Bringing in experts to review and vet online resources to ensure they are accurate and accessible that support scientists worldwide in their packaging journey.
* Running a constructive peer review process where people volunteer their time to provide feedback on a package’s usability, function, and structure.

In creating the guide, we knew we were tackling hard, complex technical problems. But also there is a tremendous amount of collective knowledge in this community that can be harnessed and focused on the problem at hand. By focusing people on the users of the guide first, rather than tools, we were able to build consensus.

We were able to move forward together.

<figure>
  <img src="{{ site.baseurl }}/images/blog/2024/june/pyopensci-python-packaging-people-can-help.png" alt="Image on a dark purple background that says - people can help. below are two people stick figure like loking at each other and smiling while chatting." style="width: 70%;" />
  <figcaption><i class="fa-solid fa-heart-pulse"></i> When tackling Python's thorniest challenges, people can help.  <i class="fa-solid fa-heart-pulse"></i>.</figcaption>
</figure>

### Create early wins for users

> As an educator, I’ve always known that the best way to empower someone new to any topic is to create a learning environment where they quickly make progress. I’ve [collected data on what approaches work better than others](https://osf.io/preprints/osf/xdj4z). I’ve explored how certain presentations of information empower, while others break people down triggering imposter syndrome and even isolation within a science field.

Reviewer: ```suggestion
As an educator, I’ve always known that the best way to empower someone new to any topic is to create a learning environment where they quickly make progress. I’ve [collected data on what approaches work better than others](https://osf.io/preprints/osf/xdj4z). I’ve explored how certain presentations of information empower, while others break people down, triggering imposter syndrome and even isolation within a science field.
```

The key is early wins.

Early wins empower users to:

* Want to do more
* Feel like they belong and can be successful learning the skill at hand
* Try again and expand their knowledge

> * Even help their colleagues which is the biggest win of all. \

Reviewer: ```suggestion
* Even help their colleagues, which is the biggest win of all. 
```

Reviewer: I don't know if I would personally put helping my colleagues above a sense of belonging!

Author: oh interesting. i definitely have found that when you run mentorship programs and or any sort of learning program / envt, the most empowering thing for a student is when you bring them back or they can get to a state where they are a  mentor  / able to help others. Because

1. it builds even more confidence 
2. you learn a ton as a mentor / helping others --> more confidence / more belonging and
3. when a new mentee / new learner sees someone "like them" that has moved into a mentorship role or a role where that person is helping others they then see themselves as belonging even more!!

the same is true for a classroom. when a student realizes they can help someone else it's really a cool confidence building situation. 

That is when i know the class has been successful - they no longer depend on the teacher /the online resources / etc. they are empowered!

i actually studied this - specifically with tribal and other underserved schools. it's powerful.

High quality documentation (or tutorials) helps users achieve this by providing quickstarts and resources that ensure they can get up and running quickly and have early positive experiences.

<figure>
  <img src="{{ site.baseurl }}/images/blog/2024/june/pyopensci-python-packaging-create-early-wins.png" alt="Image on a dark purple background that says Create Early Wins in your docs. on the left is a stressed out stick figure typing at a computer. on the right is a relaxed stick figure typing away. " style="width: 70%;" />
  <figcaption><i class="fa-solid fa-heart-pulse"></i> By creating get-start workflows, tutorials and resources that users can easily reproduce and make progress using quickly, you are empowering beginners to stick with the task at hand. And you are building user confidence in the ecosystem.  <i class="fa-solid fa-heart-pulse"></i>.</figcaption>
</figure>

### Fundamentals First

> When teaching beginners, break things down to their fundamental concepts. In the case of packaging, this meant explaining pure Python packaging in its most basic form, without immediately discussing tools. Notice that the diagram below doesn’t discuss tools at all!

Reviewer: ```suggestion
When teaching beginners, break things down to their fundamental concepts. In the case of packaging, this means explaining pure Python packaging in its most basic form, without immediately discussing tools. Notice that the diagram below doesn’t discuss tools at all!
```

<figure>
  <img src="{{ site.baseurl }}/images/blog/2024/june/pyopensci-python-packaging-tmo.png" alt="Dark purple background image that says no TMO. Below is a yellow lab retriever with a purple ice bag on it's head looking sad and directly at you." style="width: 100%;" />
  <figcaption><i class="fa-solid fa-heart-pulse"></i> When it comes to packaging, too many options and too many opinions get in the way of a positive user experience. We need to work towards breaking things down and providing opinionated paths for those who want and need them.  <i class="fa-solid fa-heart-pulse"></i>.</figcaption>
</figure>

### Friends don’t let friends package alone

> When my talk finished, I felt this sense of relief but also I felt love, support and deep passion for this topic and the community. It’s more than just packaging. It is about empowering people with skills that will help them tackle the biggest scientific challenges that we face as humans. Science is about data. And to process data we need tools - open tools that are accessible for scientists to learn and use. It’s about carving out space for scientists to build and use these tools. And also to make space for people who may feel like outsiders because of who they are, to participate.

Reviewer: ```suggestion
When I finished my talk, I felt this sense of relief but also I felt love, support and deep passion for this topic and the community. It’s more than just packaging. It is about empowering people with skills that will help them tackle the biggest scientific challenges that we face as humans. Science is about data. And to process data we need tools - open tools that are accessible for scientists to learn and use. It’s about carving out space for scientists to build and use these tools. And also to make space for people who may feel like outsiders, because of who they are, to participate.
```

> I’m excited to have the privilege of moving the pyOpenSci mission forward. Of getting to connect with so many amazing people. I’m excited to continue to facilitate and support better science through better software and to empower more people with skills. And maybe, through this talk, a little piece of that child-like imposter within me has been chipped away!

Reviewer: ```suggestion
I’m excited to have the privilege of moving the pyOpenSci mission forward, and of getting to connect with so many amazing people. I’m excited to continue to facilitate and support better science through better software and to empower more people with skills. And maybe, through this talk, a little piece of that child-like imposter within me has been chipped away!
```
